text/0028-scaling-codeowners.md

```diff
@@ -0,0 +1,63 @@
+
+# Goals:
+
+1. Users can upload a Code Owners File Size that's up to 3Mb in size
+2. The processing time to calculate assignments should not increase significantly. 
+
+# How it Works (simplified)
+
+We have Code Mappings that map the transformation of the source code to the production build. Code Mappings help us reverse-engineer the file paths from an error’s stacktrace to source code file paths. 
+
+A customer can apply a Code Mapping to a CODEOWNERS file to get the file paths that should match against the stacktrace file paths. The transformed output is stored in `sentry_projectcodeowners` and is what Sentry will use to determine assignments.
+
+When Sentry ingests a new event, we go through each line in our transformed CODEOWNERS and see if each line has a match across any of the frames in the stacktrace. We only use the last matching line for the assignment.
+
+# Current System Constraints
+
+- We store the original and transformed CODEOWNERS file in a `sentry_projectcodeowners` row. However to prevent overloading Postgres from queries on every event, we cache the row in Memcache. Memcache can only support 1mb. Rows larger than 1mb will hit the database directly. Since we store the original and transformed CODEOWNERS in the same row, files larger than 500k chars will exceed this limitation.
+- On each event, the assignment calculation is currently O((m*l + o)*n) time complexity, where *m* is the # of lines in the CODEOWNERS file, *l* is the number of code mappings mapped to the project, *o* is the number of lines in the Ownership Rules and n is the number of lines in the stacktrace. Essentially quadratic.
+- We recalculate the assignment for each event, even if the Ownership Rules and CODEOWNERS file has not been updated.
+
+# Proposal
+
+First we need baseline metrics. We need to know how long it currently takes a worker to calculate the assignments with a 100k file.
+
+> Memcache can only support 1mb.
+
+To handle Memcache’s 1mb limit, we can zip to compress the rows and store that. However, we need to consider the memory limits of our workers during the unzip process. We need to consider the impact of (3mb CODEOWNERS raw file  + 3mb CODEOWNERS transformed file)* x number of code mappings applied to that file for a project is potentially unbounded. Two options:
+
+- We can implement a limit on the number of codemappings per (repository, project)
+- We can run the calculation in batches
+ 
+
+> The calculation is currently O((m*l + o)*n) time complexity
+
+To handle the quadratic time complexity, we can use Aho-Corasick Algorithm to bring the time complexity down to O(n). [https://pyahocorasick.readthedocs.io/en/latest/](https://pyahocorasick.readthedocs.io/en/latest/) Open to better algorithm suggestions.
```

Comment:
This seems like the right idea. I don't like that it is using pickle to serialize to disk though, I would prefer that we figure out how to serialize to json if we want to cache this instead of calculating it every time.

Reply:
Aho-corasick seems sensible. One thing I want to point out though is that this problem is similar to the user agent parsing problem we have in Relay where we ultimately decided against building large regex sets / aho-corasick states after benchmarking this. In particular actually building this in python might result in you spending more time building the aho-corasick internal states every time you match them than actually doing basic matches.

Reply:
I would probably actually try to do the parsing and see if it is indeed faster. The best speedups we got on mass matching btw was using the Rust regex library which runs circles around the python one. We are currently using this already for the glob matching in Python which calls into a Rust implementation via the relay library.

Reply:
One thing we should consider as well is that for codeowner files, most rules are exact matches or doing suffix matching, and look something like
`path/to/my/file.py`
`path/to/my/*`
and relatively few looks like
`*/to/my/file.py`
`path/*/my/file.py`

It might be simpler to build a relatively simple trie here that handles the majority of rules, and then just run the regexes for the rules that are the exception here. 
```
path
       \
       to
        \
         my
          /\
         *  file.py
```
If we hit a `*` in the trie then we just instantly pass, and then same thing if we hit a leaf node.

Reply:
How big would the Trie for all projects be in memory? You could consider running this part of code owners detection in a dedicated celery task and load everything in memory at startup (not per event) and avoid the fetch of up to 1Mb from memcache per event. 
We can run Kubernetes pods with 10s of Gb of memory

Reply:
@NisanthanNanthakumar couple notes from our in person convo:
1. likely need to cache the regex values, as rust regex complation is very slow https://github.com/rust-lang/regex/blob/master/PERFORMANCE.md#thou-shalt-not-compile-regular-expressions-in-a-loop
2. definitely test rust code directly, see if python binding potentially could make it slow